src/Microsoft.Framework.Logging/LoggerExtensions.cs

```diff
@@ -18,6 +18,131 @@ public static class LoggerExtensions
         private static readonly Func<object, Exception, string> _loggerStructureFormatter = (state, error)
             => LoggerStructureFormatter((ILoggerStructure)state, error);
 
+        private static readonly Func<object, Exception, bool, string> SecureMessage =
+            (messages, error, secure) =>
+            {
+                if (secure)
+                {
+                    return ((string[])messages)[0];
+                }
+                return ((string[])messages)[1];
+            };
+
+        private static readonly Func<object, Exception, bool, string> SecureMessageAndError =
+            (messages, error, secure) =>
+            {
+                if (secure)
+                {
+                    return string.Format(CultureInfo.CurrentCulture, "{0}{1}{2}", ((string[])messages)[0], Environment.NewLine, error);
+                }
+                return string.Format(CultureInfo.CurrentCulture, "{0}{1}{2}", ((string[])messages)[1], Environment.NewLine, error);
+            };
+
+        private static readonly Func<object, Exception, bool, string> _secureLoggerStructureFormatter =
+            (state, error, secure) =>
+            {
+                var formatter = (ILoggerStructure[]) state;
+
+                if (secure)
+                {
+                    return LoggerStructureFormatter(formatter[0], error);
+                }
+                return LoggerStructureFormatter(formatter[1], error);
+                
```

Comment:
Need to clean up lots of formatting (extra newline here; lots of other places are missing spaces).